text/000-sfc-and-template-import-primitives.md

```diff
@@ -0,0 +1,875 @@
+import { setComponentTemplate } from '@ember/template-factory';
+
+const template = /* embeddable template here */
+
+class MyComponent extends Component {
+  /* ... */
+}
+
+setComponentTemplate(template, MyComponent);
+```
+
+For convenience, `setComponentTemplate` returns the same component class that
+was passed, allowing the class to be defined and associated with a template
+in a single expression:
+
+```js
+export default setComponentTemplate(template, class MyComponent extends Component {
+  /* ... */
+});
+```
+
+Even though this is a runtime API, template association is intended to be as
+static and restricted as possible, allowing us to enable better optimizations
+in the future:
+
+* Templates should be set on component classes as early as possible, ideally
+  immediately after the class is defined.
+* A component's template cannot be set more than once. Attempting to change a
+  component's template after one has been set will throw an exception.
+* Setting a component's template after rendering has begun (even if no previous
+  template was set) also throws an exception.
+* Every instance of a given component has the same template. It is not
+  possible to dynamically change a component's template at runtime.
+
+Component subclasses inherit their template from the nearest superclass that
+has had a template set via `setComponentTemplate`. If no superclass has an
+associated template, Ember will fall back to the current template resolution
+rules.
+
+Templates set via `setComponentTemplate`, including those set on a
+superclass, take precedence over the `layout` property, container lookup, or
+any other existing mechanism of template resolution. For example, a template
+set on a superclass via `setComponentTemplate` would take precedence over a
+subclass's `layout` property (if it had one). However, a template set via
+`setComponentTemplate` on the subclass would take precedence over the
+superclass.
+
+### Template Scope
+
+In this context, _scope_ refers to the set of names that can be used to refer
+to values, helpers, and other components from within your template. For
+example, when you type `{{@firstName}}`, `{{this.count}}`, or `<UserAvatar
+/>`, how does Ember know what each of these refers to? The answer to that is
+determined by the template's scope.
+
+Today, the names available inside a particular template are determined by a
+few different rules:
+
+1. `this` always refers to the component instance (if the template has a
+   backing class).
+2. Arguments (like `@firstName`) are placed into scope when you invoke the
+   component.
+3. Block parameters (like `item` in `{{#each @items as |item|}}`) are in
+   scope, but only inside the block.
+4. Anything else, like components and helper names, must be looked up
+   dynamically through Ember's container. This lookup process is somewhat
+   complex and not always as fast as we'd like.
+
+Because the rules for how components and helpers are looked up are implicit
```

Comment:
I find this ember's least attractive property. I'm glad it's called out explicitly here